docs/source/tutorials/qat_finetune.rst

```diff
@@ -0,0 +1,371 @@
+.. _qat_finetune_label:
+
+===========================
+Finetuning Llama3 with QAT
+===========================
+
+Quantization-Aware Training (QAT) is a common technique for users to quantize their
+models without incurring significant degradations in accuracy or perplexity. In this
+tutorial, we’ll walk through how to apply QAT during fine-tuning, quantize the
+resulting model, and evaluate your quantized model using torchtune.
+
+.. grid:: 2
+
+    .. grid-item-card:: :octicon:`mortar-board;1em;` What you will learn
+
+      * What QAT is and how it helps reduce quantization degradation
+      * How to run QAT during fine-tuning in torchtune
+      * End-to-end example of connecting QAT, quantization, and evaluation recipes
+
+    .. grid-item-card:: :octicon:`list-unordered;1em;` Prerequisites
+
+      * Be familiar with :ref:`torchtune<overview_label>`
+      * Make sure to :ref:`install torchtune<install_label>`
+      * Make sure you have downloaded the :ref:`Llama3-8B model weights<download_llama_label>`
+
+What is QAT?
+------------
+
+Quantization-Aware Training (QAT) refers to simulating quantization numerics during
+training or fine-tuning, with the end goal of ultimately producing a higher quality
+quantized model compared to simple post-training quantization (PTQ). During QAT,
+the weights and/or activations are “fake quantized”, meaning they are transformed
+as if they were being quantized, but kept in the original data type (e.g. bfloat16)
+without being actually cast to lower bit-widths. Thus, fake quantization allows the
+model to adjust for quantization noise when updating the weights, hence the training
+process is “aware” that the model will ultimately be quantized after training.
+
+.. code-block:: python
+
+  # PTQ: x_q is quantized and cast to int8
+  # scale and zp refer to parameters used to quantize x_float
+  # qmin and qmax refer to the range of quantized values
+  x_q = (x_float / scale + zp).round().clamp(qmin, qmax).cast(int8)
+
+  # QAT: x_fq is still in float
+  # Fake quantize simulates the numerics of quantize + dequantize
+  x_fq = (x_float / scale + zp).round().clamp(qmin, qmax)
+  x_fq = (x_fq - zp) * scale
+
+QAT typically involves applying a transformation to your model before and after training.
+For example, in the `torchao QAT implementation <https://github.com/pytorch/ao/blob/v0.2.0/torchao/quantization/prototype/qat.py>`_,
+these are represented as the prepare and convert steps: (1) prepare inserts fake quantize
+operations into linear layers, and (2) convert transforms the fake quantize operations
+to actual quantize operations after training, thereby producing a quantized model.
+Between these two steps, training can proceed exactly as before.
+
+.. image:: /_static/img/qat_diagram.png
```

Comment:
Nice diagram! One comment -- as a noob reader it's not immediately clear from the preceding text why the converted version does dequant on Wq (maybe this is obvious for someone familiar with quantization algos but idk how many readers of this tutorial that'll apply to). Might be good to add a sentence explaining forward logic in the quantized model.

Reply:
Yeah this diagram contains some implementation details. I simplified it a bit and added a sentence to explain